File: src/main/java/de/jepfa/obfusser/util/encrypt/hints/EncryptedHintChar.java
package de.jepfa.obfusser.util.encrypt.hints;

public class EncryptedHintChar extends HintChar {

    private static final int MAX_ROUND_TRIPS = 3;

    private int roundTrips;


    public static EncryptedHintChar ofDecrypted(char c) {
        return new EncryptedHintChar(c, 0);
    }

    /**
     *
     * @param encHintData format must be char + roundtrip (one digit), e.g. "x2"
     * @return
     */
    public static EncryptedHintChar ofEncrypted(String encHintData) {
        char[] chars = encHintData.toCharArray();
        char encHint = chars[0];
        int roundTrips = Integer.parseInt(String.valueOf(chars[1]));
        return new EncryptedHintChar(encHint, roundTrips);
    }

    public EncryptedHintChar(char hint, int roundTrips) {
        super(hint);
        this.roundTrips = roundTrips;
    }

    public int getRoundTrips() {
        return roundTrips;
    }

    public boolean doNext() {
        roundTrips++;
        return isGoNext() && getRoundTrips() < MAX_ROUND_TRIPS;
    }

    /**
     * @return format is char + roundtrip (one digit), e.g. "x2"
     */
    public String getHintStoreString() {
        return String.valueOf(getHint()) + roundTrips;
    }

    public void apply(HintChar hintChar) {
        if (hintChar != null) {
            setHint(hintChar.getHint());
            setGoNext(hintChar.isGoNext());
        }
    }

    @Override
    public String toString() {
        final StringBuffer sb = new StringBuffer("EncryptedHintChar{");
        sb.append("hint='").append(getHint()).append('\'');
        sb.append(", goNext=").append(isGoNext());
        sb.append(", roundTrips=").append(roundTrips);
        sb.append('}');
        return sb.toString();
    }

}

File: src/main/java/de/jepfa/obfusser/util/encrypt/hints/EncryptedHintChar.kt
package de.jepfa.obfusser.util.encrypt.hints

class EncryptedHintChar(hint: Char, roundTrips: Int) : HintChar(hint) {

    var roundTrips: Int = 0
        private set

    /**
     * @return format is char + roundtrip (one digit), e.g. "x2"
     */
    val hintStoreString: String
        get() = hint.toString() + roundTrips

    init {
        this.roundTrips = roundTrips
    }

    fun doNext(): Boolean {
        roundTrips++
        return isGoNext && roundTrips < MAX_ROUND_TRIPS
    }

    fun apply(hintChar: HintChar?) {
        hintChar?.let {
            hint = hintChar.hint
            isGoNext = hintChar.isGoNext
        }
    }

    override fun toString(): String {
        val sb = StringBuffer("EncryptedHintChar{")
        sb.append("hint='").append(hint).append('\'')
        sb.append(", goNext=").append(isGoNext)
        sb.append(", roundTrips=").append(roundTrips)
        sb.append('}')
        return sb.toString()
    }

    companion object {

        private val MAX_ROUND_TRIPS = 3


        fun ofDecrypted(c: Char): EncryptedHintChar {
            return EncryptedHintChar(c, 0)
        }

        /**
         *
         * @param encHintData format must be char + roundtrip (one digit), e.g. "x2"
         * @return
         */
        fun ofEncrypted(encHintData: String): EncryptedHintChar {
            val chars = encHintData.toCharArray()
            val encHint = chars[0]
            val roundTrips = Integer.parseInt(chars[1].toString())
            return EncryptedHintChar(encHint, roundTrips)
        }
    }

}
